compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1648,18 +1648,35 @@ object messages {
     val explanation = "Method inlining prohibits calling superclass methods, as it may lead to confusion about which super is being called."
   }
 
-  case class ModifiersNotAllowed(flags: FlagSet, sort: String)(implicit ctx: Context)
+  case class ModifiersNotAllowed(flags: FlagSet, printableType: Option[String])(implicit ctx: Context)
     extends Message(ModifiersNotAllowedID) {
     val kind = "Syntax"
-    val msg = s"modifier(s) `$flags' not allowed for $sort"
+    val msg = s"modifier(s) `$flags' not allowed for ${printableType.getOrElse("combination")}"
     val explanation = {
-      val code = "sealed def y: Int = 1"
+      val first = "sealed def y: Int = 1"
+      val second = "sealed lazy class z"
       hl"""You tried to use a modifier that is inapplicable for the type of item under modification
          |
+         |  Here is a list of modifiers and applicable item types
+         |
+         |
+         |  Modifier          ::=  LocalModifier
+         |                         AccessModifier
+         |                         `override'         traits, methods
+         |  LocalModifier     ::=  `abstract'         classes, traits
+         |                         `final'            classes, traits, methods
+         |                         `sealed'           classes, traits
+         |                         `implicit'         variables, methods, classes
+         |                         `lazy'             values
+         |  AccessModifier    ::=  (`private' | `protected') [AccessQualifier] methods, classes
+         |  AccessQualifier   ::=  `[' (id | `this') `]'
+         |
```

Comment:
Having a copy of the syntax here might be nice, but it'd be better if we could link to the spec (we could open an issue for this instead).